src/DurableSDK/PowerShellServices.cs

```diff
@@ -6,50 +6,146 @@
         }
 
-        public void SetOrchestrationContext(OrchestrationContext orchestrationContext)
+        public OrchestrationBindingInfo SetOrchestrationContext(
+            ParameterBinding context,
+            out IExternalOrchestrationInvoker externalInvoker)
         {
-            _pwsh.AddCommand(SetFunctionInvocationContextCommand)
-                .AddParameter("OrchestrationContext", orchestrationContext)
-                .InvokeAndClearCommands();
+            externalInvoker = null;
+            OrchestrationBindingInfo orchestrationBindingInfo = new OrchestrationBindingInfo(
+                context.Name,
+                JsonConvert.DeserializeObject<OrchestrationContext>(context.Data.String));
+
+            if (hasExternalDurableSDK)
+            {
+                Collection<Func<PowerShell, object>> output = _pwsh.AddCommand(SetFunctionInvocationContextCommand)
+                    // The external SetFunctionInvocationContextCommand expects a .json string to deserialize
+                    // and writes an invoker function to the output pipeline.
+                    .AddParameter("OrchestrationContext", context.Data.String)
+                    .InvokeAndClearCommands<Func<PowerShell, object>>();
+
+                // If more than 1 element is present in the output pipeline, we cannot trust that we have
+                // obtained the external orchestrator invoker; i.e the output contract is not met.
+                var outputContractIsMet = output.Count() == 1;
+                if (outputContractIsMet)
+                {
+                    externalInvoker = new ExternalInvoker(output[0]);
+                }
+                else
+                {
+                    throw new InvalidOperationException($"Only a single output was expected for an invocation of {SetFunctionInvocationContextCommand}");
+                }
+            }
+            else
+            {
+                _pwsh.AddCommand(SetFunctionInvocationContextCommand)
+                    .AddParameter("OrchestrationContext", orchestrationBindingInfo.Context)
+                    .InvokeAndClearCommands();
+            }
+            hasInitializedDurableFunctions = true;
+            return orchestrationBindingInfo;
+        }
+        
 
-            _hasSetOrchestrationContext = true;
+        public void AddParameter(string name, object value)
+        {
+            _pwsh.AddParameter(name, value);
         }
 
         public void ClearOrchestrationContext()
         {
-            if (_hasSetOrchestrationContext)
+            if (hasInitializedDurableFunctions)
             {
                 _pwsh.AddCommand(SetFunctionInvocationContextCommand)
                     .AddParameter("Clear", true)
                     .InvokeAndClearCommands();
             }
         }
 
+        public void TracePipelineObject()
+        {
+            _pwsh.AddCommand("Microsoft.Azure.Functions.PowerShellWorker\\Trace-PipelineObject");
```

Comment:
Should we create a file with constants (for the cmdlet names) which can be used across the whole project?

Reply:
Good idea! Done